packages/framework/src/client/resource-request.js

```diff
@@ -0,0 +1,111 @@
+import debugFactory from 'debug';
+import { isDateEarlier } from '../utils/dates';
+import { SECOND } from '../utils/constants';
+
+const DEFAULT_TIMEOUT = 30 * SECOND;
+
+export const STATUS = {
+	failed: 'Failed',
+	inFlight: 'In Flight',
+	complete: 'Complete',
+	overdue: 'Overdue',
+	scheduled: 'Scheduled',
+	timedOut: 'Timed Out',
+	unnecessary: 'Unnecessary',
+};
+
+export default class ResourceRequest {
+	constructor( resourceName, requirement, resourceState, now ) {
+		this.debug = debugFactory( 'fresh-data:request(' + resourceName + ')' );
+		this.resourceName = resourceName;
+		this.time = calculateRequestTime( requirement, resourceState, now );
+		this.timeout = requirement.timeout || DEFAULT_TIMEOUT;
+		this.promise = null;
+		this.timeRequested = null;
+		this.timeCompleted = null;
+		this.error = null;
+
+		if ( this.time ) {
+			const seconds = ( this.time.getTime() - now.getTime() ) / 1000.0;
+			this.debug( `New request for "${ this.resourceName }" to be fetched in ${ seconds }s` );
+		}
+	}
+
+	addRequirement( requirement, resourceState, now ) {
+		const status = this.getStatus( now );
+		if ( STATUS.scheduled === status || STATUS.overdue === status ) {
+			const requestTime = calculateRequestTime( requirement, resourceState, now );
+			if ( isDateEarlier( this.time, requestTime ) ) {
+				this.time = requestTime;
+				const seconds = ( this.time.getTime() - now.getTime() ) / 1000.0;
+				this.debug( `Rescheduling request for "${ this.resourceName }" to fetched in ${ seconds }s` );
+			}
+			this.timeout = Math.min( this.timeout, requirement.timeout || DEFAULT_TIMEOUT );
+			return true;
+		}
+		this.debug( `Cannot add requirement to request with "${ this.getStatus( now ) }" status` );
+		return false;
+	}
+
+	getStatus = ( now = new Date() ) => {
+		if ( ! this.time ) {
+			return STATUS.unnecessary;
+		}
+		if ( this.timeRequested ) {
+			if ( this.timeCompleted ) {
+				if ( this.error ) {
+					return STATUS.failed;
+				}
+				return STATUS.complete;
+			}
+			if ( now.getTime() - this.timeRequested > this.timeout ) {
+				return STATUS.timedOut;
+			}
+			return STATUS.inFlight;
+		}
+		if ( isDateEarlier( now, this.time ) ) {
+			return STATUS.overdue;
+		}
+		return STATUS.scheduled;
+	}
+
+	getTimeLeft = ( now = new Date() ) => {
+		return this.time.getTime() - now.getTime();
+	}
+
+	requested = ( promise, now = new Date() ) => {
```

Comment:
Maybe `submitRequest` or `startRequest` would work here?

Reply:
Hmm, in this case the actual request has to happen outside of the ResourceRequest object because multiple ResourceRequests may be fulfilled by the same operation/promise. i.e. one operation may batch multiple ResourceRequests. In this case, a singular operation promise is given to multiple requests for their own tracking. The purpose of this is so the ResourceRequest object can always report an accurate status without having to have supporting code iterate through all ResourceRequests to set their states for them (like when a promise is fulfilled or rejected)